components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/DigestAuthMediator.java

```diff
@@ -0,0 +1,511 @@
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.wso2.carbon.apimgt.gateway.mediators;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpHeaders;
+import org.apache.synapse.ManagedLifecycle;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.SynapseEnvironment;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.mediators.AbstractMediator;
+
+import java.net.URI;
+import java.util.*;
+
+/**
+ * This mediator would set the authorization header of the request that is to be sent to the endpoint,
+ * making use of the 401 Unauthorized response received from the first request.
+ */
+
+public class DigestAuthMediator extends AbstractMediator implements ManagedLifecycle {
+
+    private static final Log log = LogFactory.getLog(DigestAuthMediator.class);
+
+    private static final String POSTFIX = "POSTFIX";
+    private static final String HTTP_METHOD = "HTTP_METHOD";
+    private static final String MESSAGE_BODY = "MessageBody";
+    private static final String UNAMEPASSWORD = "UNAMEPASSWORD";
+    private static final String NONCE_COUNT = "NonceCount";
+    private static final String INIT_NONCE_COUNT = "00000000";
+    private static final String AUTH_HEADER = "AuthHeader";
+    private static final String BACKEND_URL = "BACKEND_URL";
+    private static final String CHARSET = "UTF-8";
+
+    /**
+     * This method is used to split the WWW-Authenticate header sent from the backend
+     *
+     * @param wwwHeaderSplits The array containing 2 Strings, where the first is an empty String always and the next is
+     *                        the String containing the digest challenge sent in the WWW-Authenticate header
+     * @return A String array containing all the required parameters for the Authorization header processing
+     */
+    public String[] splitDigestHeader(String[] wwwHeaderSplits) {
+
+        String realm = null;
+        String qop = null;
+        String serverNonce = null;
+        String opaque = null;
+        String algorithm = null;
+
+        String wwwHeaderValueString = wwwHeaderSplits[1];
+
+        //Should split from ", " instead of "," because this might fail when qop or algorithm has multiple values.
+        String wwwHeaderValueArray[] = wwwHeaderValueString.split(", ");
+
+        for (String keyval : wwwHeaderValueArray) {
+
+            String key = keyval.substring(0, keyval.indexOf("="));
+            String value = keyval.substring(keyval.indexOf("=") + 1);
+
+            if ("realm".equals(key.trim())) {
```

Comment:
You can move realm, qop, nonce, opaque, algorithm etc etc to APIConstants class and reuse in your code